service/history/historyReplicator.go

```diff
@@ -185,8 +154,160 @@ func (r *historyReplicator) ApplyEvents(request *h.ReplicateEventsRequest) error
+
+		case shared.EventTypeWorkflowExecutionFailed:
+			msBuilder.ReplicateWorkflowExecutionFailedEvent(event)
+
+		case shared.EventTypeWorkflowExecutionTimedOut:
+			msBuilder.ReplicateWorkflowExecutionTimedoutEvent(event)
+
+		case shared.EventTypeWorkflowExecutionCanceled:
+			msBuilder.ReplicateWorkflowExecutionCanceledEvent(event)
+
+		case shared.EventTypeWorkflowExecutionTerminated:
+			msBuilder.ReplicateWorkflowExecutionTerminatedEvent(event)
+
+		case shared.EventTypeWorkflowExecutionContinuedAsNew:
+			// ContinuedAsNew event also has history for first 2 events for next run as they are created transactionally
+			newRunHistory := request.NewRunHistory
+			startedEvent := newRunHistory.Events[0]
+			startedAttributes := startedEvent.WorkflowExecutionStartedEventAttributes
+			dtScheduledEvent := newRunHistory.Events[1]
+			domainEntry, err := r.shard.GetDomainCache().GetDomainByID(domainID)
+			if err != nil {
+				return err
+			}
+			domainName := domainEntry.GetInfo().Name
+
+			// History event only have the parentDomainName.  Lookup the domain ID from cache
+			var parentDomainID *string
+			if startedAttributes.ParentWorkflowDomain != nil {
+				parentDomainEntry, err := r.shard.GetDomainCache().GetDomain(startedAttributes.GetParentWorkflowDomain())
+				if err != nil {
+					return err
+				}
+				parentDomainID = &parentDomainEntry.GetInfo().ID
+			}
+
+			newRunID := event.WorkflowExecutionContinuedAsNewEventAttributes.GetNewExecutionRunId()
+
+			newExecution := shared.WorkflowExecution{
+				WorkflowId: request.WorkflowExecution.WorkflowId,
+				RunId:      common.StringPtr(newRunID),
+			}
+
+			// Create mutable state updates for the new run
+			newStateBuilder := newMutableStateBuilder(r.shard.GetConfig(), r.logger)
+			newStateBuilder.ReplicateWorkflowExecutionStartedEvent(domainID, parentDomainID, newExecution, uuid.New(),
+				startedAttributes)
+			di := newStateBuilder.ReplicateDecisionTaskScheduledEvent(dtScheduledEvent.GetEventId(),
+				dtScheduledEvent.DecisionTaskScheduledEventAttributes.TaskList.GetName(),
+				dtScheduledEvent.DecisionTaskScheduledEventAttributes.GetStartToCloseTimeoutSeconds())
+			nextEventID := di.ScheduleID + 1
+			newStateBuilder.executionInfo.NextEventID = nextEventID
+			newStateBuilder.executionInfo.LastFirstEventID = startedEvent.GetEventId()
+			newStateBuilder.ApplyReplicationStateUpdates(request.GetVersion(), di.ScheduleID)
+			// Set the history from replication task on the newStateBuilder
+			newStateBuilder.hBuilder = newHistoryBuilderFromEvents(newRunHistory.Events, r.logger)
+
+			msBuilder.ReplicateWorkflowExecutionContinuedAsNewEvent(domainID, domainName, event, startedEvent, di,
+				newStateBuilder)
+
+			// Generate a transaction ID for appending events to history
+			transactionID, err := r.shard.GetNextTransferTaskID()
+			if err != nil {
+				return err
+			}
+			err = context.replicateContinueAsNewWorkflowExecution(newStateBuilder, nil, nil, transactionID)
```

Comment:
for the new run, you don't need timer tasks and transfer tasks?

Reply:
Support for generating Timer and Transfer task for replicated event will come later.  This change is only about generating replication tasks.